tests/unittests/test_runs/test_fallback_datasource.py

```diff
@@ -0,0 +1,240 @@
+        self.ds = dscd.DataSourceConfigDrive
+
+    def cloud_init(self, initer):
+        initer.read_cfg()
+        initer.initialize()
+        initer.fetch()
+        iid = initer.instancify()
+        initer.update()
+        return iid
+
+    def remove_seed_dir(self, seed_dir):
+        self.patched_funcs.close()
+        util.del_dir(seed_dir)
+        util.del_file(os.path.join(self.tmp, "run/cloud-init/.instance-id"))
+        util.del_file(os.path.join(self.tmp, ".instance-id"))
+
+    def test_invalid_config_drive(self):
+        seed_dir = os.path.join(self.paths.seed_dir, "config_drive")
+        populate_dir(seed_dir, CFG_DRIVE_FILES_V2)
+
+        sys_cfg = {
+            'datasource_list': ['ConfigDrive'],
+            'datasource': {'ConfigDrive': {'dsmode': 'local'}}
+        }
+
+        cloud_cfg = safeyaml.dumps(sys_cfg)
+        util.ensure_dir(os.path.join(self.tmp, 'etc', 'cloud'))
+        util.write_file(os.path.join(self.tmp, 'etc',
+                                     'cloud', 'cloud.cfg'), cloud_cfg)
+        dsrc = self.ds(sys_cfg=sys_cfg, distro=None, paths=self.paths)
+        ret = dsrc.get_data()
+        self.assertEqual(
+            dsrc.subplatform, 'seed-dir (%s)' % seed_dir)
+        self.assertTrue(ret)
+        # simulate first boot with config drive
+        self.reRoot(self.tmp)
+        initer = stages.Init()
+        initer.datasource = dsrc
+        iid_with_dsrc = self.cloud_init(initer)
+        # simulate subsequent boot without config drive
+        self.remove_seed_dir(seed_dir)
+        self.reRoot(self.tmp)
+        initer = stages.Init(ds_deps=['FILESYSTEM'])
+        iid_without_dsrc = self.cloud_init(initer)
+        self.assertEqual(iid_with_dsrc, iid_without_dsrc)
+
+    def test_invalid_openstack_uri(self):
+        _register_uris(self.VERSION, {}, {}, OS_FILES)
+        sys_cfg = {
+            'datasource_list': ['OpenStack'],
+            'datasource': {'OpenStack': {'dsmode': 'net'}}
+        }
+        cloud_cfg = safeyaml.dumps(sys_cfg)
+        util.ensure_dir(os.path.join(self.tmp, 'etc', 'cloud'))
+        util.write_file(os.path.join(self.tmp, 'etc',
+                                     'cloud', 'cloud.cfg'), cloud_cfg)
+        self.ds = dsos.DataSourceOpenStack
+        ds_os = self.ds(sys_cfg=sys_cfg, distro=None, paths=self.paths)
+        MOCK_PATH = 'cloudinit.sources.DataSourceOpenStack.'
+        mock_path = MOCK_PATH + 'detect_openstack'
+        with tests_helpers.mock.patch(mock_path) as m_detect_os:
+            m_detect_os.return_value = True
+            found = ds_os.get_data()
+        self.assertTrue(found)
+        self.reRoot(self.tmp)
+        # simulate first boot with datasource openstack
+        initer = stages.Init()
+        initer.datasource = ds_os
+        iid_with_dsrc = self.cloud_init(initer)
+
+        # simulate subsequent boot without datasource openstack
+        util.del_file(os.path.join(self.tmp, "run/cloud-init/.instance-id"))
```

Comment:
Why do you delete the .instance-id file in this case, but not the ConfigDrive test?

Reply:
I think it was getting deleted in case of ConfigDrive also in remove_seed_dir function earlier. Now changed it latest commit

Reply:
I see, I think if you moved the del file into a def reboot(): method, then in each test

```
# simulate first boot with datasource ...
iid_with_dsrc = self.simulate_cloudinit_boot(...)

self.reboot()

# simulate datasource failure
<whatever is needed>

# do second boot

self.assertEqual()
```

Reply:
done